cmd/release-controller/upgrades.go

```diff
@@ -345,3 +349,96 @@ func syncGraphToSecret(graph *UpgradeGraph, update bool, secretClient kv1core.Se
 		glog.V(2).Infof("Saved upgrade graph state to %s/%s", ns, name)
 	}, 5*time.Minute, stopCh)
 }
+
+func releaseTagTimestamp(v string) (time.Time, error) {
+	var t time.Time
+	version, err := semverParseTolerant(v)
+	if err != nil {
+		return t, err
+	}
+	if len(version.Pre) == 0 {
+		return t, fmt.Errorf("tag %s has no timestamp", v)
+	}
+	prereleaseLast := version.Pre[len(version.Pre) - 1].VersionStr
```

Comment:
This is questionable - what are you actually trying to get here?  `len(version.Pre) - 1` is nothing that is obvious from the code.

Reply:
I'm parsing the release tag name for the timestamp. I can skip this and try to look up timestamps on the annotations on release, but I'm not sure if won't prune the stable edges in case they are missing in the annotations.